Simplify particle transfer during refinement/serialization

gassmoeller opened the pull request:
The way particles are transferred during mesh refinement and serialization at the moment is in essence identical to how DoFHandler and SolutionTransfer do the same to solution vectors. However, the function names and how they are used is completely different, and not well documented for the particle handler. This PR deprecates the old function names and introduces new functions modeled after the SolutionTransfer class. It also adds more documentation to the corresponding functions in the particle handler class to describe how they are used, and updates all examples and tests that use them.

Main changes:
- to transfer particles across a refinement instead of connecting `register_store/load_callback_function` to the triangulation signals, you now call `prepare_for_coarsening_and_refinement` before and `transfer_after_coarsening_and_refinement` after the refinement. Just like we expect SolutionTransfer to work. The old way will still work if users want to keep it, but it is not advertised anymore, because it fails for mappings like MappingQCache and MappingEulerian.
- I reworked the transfer functions so that they throw exceptions when called with triangulations that are not supported for refinement. Before #12434 we had undefined behavior in that case, after #12434 particles would just disappear. I think throwing an exception is the right thing to do, until we have a working transfer for those cases (I have a WIP, but will handle that in a separate PR)
- I kept one special case and that is, if you create a particle handler, and change its connected triangulation (shared or distributed) before adding particles, the particle handler will not complain and resize its internal container accordingly without needing calls to the `prepare...` and `transfer...` functions. In other words you can use a particle handler with serial and shared triangulations, as long as you never do refinement while particles exist. This will keep step-19 working and is also important for other cases that change the triangulation after creating the particle handler, but before creating particles, like @blaisb does in https://github.com/lethe-cfd/lethe.
- Keep the connections between triangulation and particle handler stored, to be able to disconnect them and avoid duplicating them when copying particle handlers.

I am open for suggestions about the names of the functions, or the general behavior.

marcfehling commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -718,20 +718,81 @@ namespace Particles
+     * SolutionTransfer class transfer the vector to the new mesh.
      */
     void
+    transfer_after_coarsening_and_refinement();
```
For the `CellDataTransfer` class, we use the name [`unpack`](https://www.dealii.org/developer/doxygen/deal.II/classparallel_1_1distributed_1_1CellDataTransfer.html#a18bbf81e37b71e43fe827cbc45ef4101) for this function.

We now have three different functions (interpolate, unpack, transfer_after_coarsening_and_refinement) in separate classes that do essentially the same. Should we think of a common name for it?

gassmoeller replied:
I agree, but they may be used in many places, so we should consider if it is worth it to rename all of them. I am not sure how many places use CellDataTransfer, but if we change the interface of SolutionTransfer, I think we have a lot of work and convincing to do :-), I would be curious about opinions though. 

Starting from a clean slate I think I would like the combination `prepare_for_mesh_transfer`/`unpack_after_mesh_transfer` best. `mesh_transfer` also captures the repartition case (currently we need to prepare for coarsening and refinement even if we only do a distributed repartition), and also references the name of the classes (CellDataTransfer/SolutionTransfer). And adding `_after_mesh_transfer` to the unpack function may not be necessary for CellDataTransfer, because its only purpose is the transfer, but I think `particle_handler.unpack()` is not quite descriptive enough, while `particle_handler.unpack_after_mesh_transfer()` make immediate sense. Also I like the symmetry ^^.

If you think those make sense, I am happy to go ahead for the particle handler. Alternatively, I am also ok with `prepare_for_coarsening_and_refinement`/`unpack` to at least make CellDataTransfer and ParticleHandler identical.

marcfehling commented on examples/step-70/step-70.cc:
```diff
@@ -1095,11 +1095,12 @@ namespace Step70
     fluid_tria.signals.post_distributed_refinement.connect([&]() {
-      tracer_particle_handler.register_load_callback_function(false);
+      tracer_particle_handler.transfer_after_coarsening_and_refinement();
     });
```
I think it would make sense to call these functions embedded around `fluid_tria.execute_coarsening_and_refinement()` similar to `step-68`, rather than connecting them to signals.

marcfehling commented on examples/step-70/step-70.cc:
```diff
@@ -1188,11 +1189,13 @@ namespace Step70
-      [&]() { solid_particle_handler.register_store_callback_function(); });
+    fluid_tria.signals.pre_distributed_refinement.connect([&]() {
+      solid_particle_handler.prepare_for_coarsening_and_refinement();
+    });
 
-    fluid_tria.signals.post_distributed_refinement.connect(
-      [&]() { solid_particle_handler.register_load_callback_function(false); });
+    fluid_tria.signals.post_distributed_refinement.connect([&]() {
+      solid_particle_handler.transfer_after_coarsening_and_refinement();
+    });
```
Same here.

marcfehling commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -718,20 +718,81 @@ namespace Particles
+    void
+    deserialize();
+
+    /**
+     * @deprecated: Please use prepare_for_coarsening_and_refinement() or
+     * prepare_for_serialization() instead. See there for further information
+     * about the purpose of this function.
+     */
+    DEAL_II_DEPRECATED
+    void
     register_store_callback_function();
```
Could you mark this function `DEAL_II_DEPRECATED_EARLY` according to this [changelog](https://github.com/dealii/dealii/blob/787280118173e892de6a255343638a988a24d806/doc/news/9.2.0-vs-9.3.0.h#L1131-L1144) entry.

Also, feel free to keep the original documentation of this function before the `@deprecated` tag.

marcfehling commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -718,20 +718,81 @@ namespace Particles
-     * Callback function that should be called after every refinement
-     * and after resuming from a checkpoint.  This function is used to
-     * register load_particles() with the triangulation. This function
-     * is used in step-70.
+     * @deprecated: Please use transfer_after_coarsening_and_refinement() or
+     * deserialize() instead. See there for further information about the
+     * purpose of this function.
      */
+    DEAL_II_DEPRECATED
     void
     register_load_callback_function(const bool serialization);
```
Same here.

marcfehling commented on source/particles/particle_handler.cc:
```diff
@@ -1845,76 +1833,195 @@ namespace Particles
+
+  template <int dim, int spacedim>
+  void
+  ParticleHandler<dim, spacedim>::register_load_callback(
```
I find this function name misleading. You don't register a callback function here, but actually perform the unpacking by calling `notify_ready_to_unpack`.

Maybe call it `{call|perform|trigger|activate}_load_callback` instead?

marcfehling commented on tests/serialization/particle_handler_02.cc:
```diff
@@ -107,20 +107,12 @@ test()
   {
     boost::archive::text_oarchive oa(oss, boost::archive::no_header);
+
+    particle_handler.prepare_for_serialization();
```
I would call this function directly before the `save` function in line 131. As far as I know, this function only connects the callback function to a signal in `save`, right?

marcfehling reviewed (approved):
I like this change as it simplifies the transfer interface a lot. You'll find some minor suggestions attached.

gassmoeller commented:
I think I addressed all comments. I went with `prepare_for_coarsening_and_refinement` and `unpack_after_coarsening_and_refinement` for now, but I am ok with changing that if we agree on a better unified name. Also I unified some names of private functions with the SolutionTransfer class (e.g. `register_data_attach` and `notify_ready_to_unpack` instead of `register_store_callback` and `register_load_callback`). Let me know what you think about this version.

marcfehling commented on examples/step-70/step-70.cc:
```diff
@@ -1188,11 +1169,13 @@ namespace Step70
     // particles are correctly transferred:
-    fluid_tria.signals.pre_distributed_refinement.connect(
-      [&]() { solid_particle_handler.register_store_callback_function(); });
+    fluid_tria.signals.pre_distributed_refinement.connect([&]() {
+      solid_particle_handler.prepare_for_coarsening_and_refinement();
+    });
 
-    fluid_tria.signals.post_distributed_refinement.connect(
-      [&]() { solid_particle_handler.register_load_callback_function(false); });
+    fluid_tria.signals.post_distributed_refinement.connect([&]() {
+      solid_particle_handler.unpack_after_coarsening_and_refinement();
+    });
```
I think you don't need this anymore.

marcfehling commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -717,21 +717,89 @@ namespace Particles
+     * register pack_callback() with the triangulation. This function
      * is used in step-70.
+     *
+     * @deprecated: Please use prepare_for_coarsening_and_refinement() or
```
```suggestion
     * @deprecated Please use prepare_for_coarsening_and_refinement() or
```

marcfehling commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -717,21 +717,89 @@ namespace Particles
+     * SolutionTransfer class transfer the vector to the new mesh.
+     */
+    void
+    unpack_after_coarsening_and_refinement();
```
> I am not sure how many places use CellDataTransfer, but if we change the interface of SolutionTransfer, I think we have a lot of work and convincing to do :-)

I agree. `SolutionTransfer` is so important and widely used in the `deal.II` community, that changing the interface will let us face a lot of resistance. However, finding a common nameset will make it easier for new users.

> Starting from a clean slate I think I would like the combination prepare_for_mesh_transfer/unpack_after_mesh_transfer best. mesh_transfer also captures the repartition case (currently we need to prepare for coarsening and refinement even if we only do a distributed repartition), and also references the name of the classes (CellDataTransfer/SolutionTransfer). And adding _after_mesh_transfer to the unpack function may not be necessary for CellDataTransfer, because its only purpose is the transfer, but I think particle_handler.unpack() is not quite descriptive enough, while particle_handler.unpack_after_mesh_transfer() make immediate sense. Also I like the symmetry ^^.

I also agree with you on the generalization of the suffix. How about reducing it from `mesh_transfer` to `transfer` only? This way we would end up with functions `prepare_for_transfer` and `unpack_after_transfer`, which I like very much.

What do you think of calling the post transfer function `distribute_after_transfer` instead? I'm undecided whether `unpack` sounds too technical in that regard, and that we should rather use a different word here. However, the word `distribute` is kind of biased due to `DoFHandler::distribute_dofs()`, which might lead to confusion.

marcfehling reviewed (approved):
You need to solve the merge conflicts. Afterwards we can run the testsuite!

---

> Also I unified some names of private functions with the SolutionTransfer class.

I like it this way!

gassmoeller commented:
I rebased and fixed the issues in your comments.

> What do you think of calling the post transfer function distribute_after_transfer instead?

Yes, that's perfect. `prepare_for_transfer` and `distribute_after_transfer` should work nicely for all three cases. However, I would prefer to make that a separate / follow-up PR, because it affects the other classes as well. I would leave the current PR as is, because it works best in case we decide to not rename the other classes as well. Do you agree?

marcfehling commented:
/rebuild

blaisb commented:
I really love this PR @gassmoeller , this is such a nice change.
